src/platform/tls_openssl.c

```diff
@@ -0,0 +1,3437 @@
+    }
+    return 1;
+
+}
+
+//
+// @brief Callback to yield TLS secrets to the QUIC stack.
+//
+// This function is invoked by OpenSSL to provide traffic secrets during
+// the QUIC handshake. It installs the given secret into the msquic QUIC
+// connection, either as a read (RX) key or write (TX) key depending on
+// the direction.
+//
+// @param[in] s           Pointer to the SSL connection object.
+// @param[in] prot_level  OpenSSL encryption level of the secret.
+// @param[in] dir         Direction of the key. 1 for read (RX) key,
+//                        0 for write (TX) key.
+// @param[in] secret      Pointer to the secret to be installed.
+// @param[in] secret_len  Length of the secret.
+// @param[in] arg         Unused argument (typically NULL).
+//
+// @return 1 on success, 0 on failure.
+//
+static int QuicTlsYieldSecret(SSL *S, uint32_t ProtLevel,
+                                 int Dir,
+                                 const unsigned char *NewSecret,
+                                 size_t SecretLen, void *Arg)
+{
+    CXPLAT_TLS* TlsContext = SSL_get_app_data(S);
+    CXPLAT_TLS_PROCESS_STATE* TlsState = TlsContext->State;
+    QUIC_PACKET_KEY_TYPE KeyType = (QUIC_PACKET_KEY_TYPE)ProtLevel;
+    QUIC_STATUS Status;
+    CXPLAT_SECRET Secret;
+    struct AUX_DATA *AData = GetSslAuxData(S);
+
+    UNREFERENCED_PARAMETER(Arg);
+
+    QuicTraceLogConnVerbose(
+        OpenSslNewEncryptionSecrets,
+        TlsContext->Connection,
+        "New encryption secrets (Level = %u)",
+        ProtLevel);
+
+    CxPlatTlsNegotiatedCiphers(TlsContext, &Secret.Aead, &Secret.Hash);
+
+    //
+    // Tx/Write Secret
+    //
+    if (Dir == 1) {
+        CxPlatCopyMemory(Secret.Secret, NewSecret, SecretLen);
+        CXPLAT_DBG_ASSERT(TlsState->WriteKeys[KeyType] == NULL);
+        Status =
+            QuicPacketKeyDerive(
+                KeyType,
+                TlsContext->HkdfLabels,
+                &Secret,
+                "write Secret",
+                TRUE,
+                &TlsState->WriteKeys[KeyType]);
+        if (QUIC_FAILED(Status)) {
+            TlsContext->ResultFlags |= CXPLAT_TLS_RESULT_ERROR;
+            return -1;
+        }
+
+        if (TlsContext->IsServer && KeyType == QUIC_PACKET_KEY_0_RTT) {
+            TlsContext->ResultFlags |= CXPLAT_TLS_RESULT_EARLY_DATA_ACCEPT;
+            TlsContext->State->EarlyDataState = CXPLAT_TLS_EARLY_DATA_ACCEPTED;
+        }
+    } else  {
+        CxPlatCopyMemory(Secret.Secret, NewSecret, SecretLen);
+        CXPLAT_DBG_ASSERT(TlsState->ReadKeys[KeyType] == NULL);
+        Status =
+            QuicPacketKeyDerive(
+                KeyType,
+                TlsContext->HkdfLabels,
+                &Secret,
+                "read Secret",
+                TRUE,
+                &TlsState->ReadKeys[KeyType]);
+        if (QUIC_FAILED(Status)) {
+            TlsContext->ResultFlags |= CXPLAT_TLS_RESULT_ERROR;
+            return -1;
+        }
+
+    }
+    if (ProtLevel == 3) {
```

Comment:
Can we use a named constant instead of the bare value 3?  Or a comment explaining which level this is?